### ISSUE TITLE
Calc_loc for specifying previous OptimizeFW for a StaticFW?

### ISSUE THREAD

**Issue by blondegeek** [hackingmaterials/atomate#94]
I'm cleaning up some old MatMethods workflows to push to atomate and I noticed that the current atomate.vasp.fireworks.StaticFW no longer has keywords for calc_loc.

Within a single workflow, I run OptimizeFW for two structures in parallel followed by a StaticFW for each structure (these structures are then used together later to create interpolated structures). I use calc_loc to specify which of the OptimizeFW runs the StaticFW should use for copy_vasp_outputs.

It's my understanding that calc_loc = True will grab the location of the most recently completed calculation (rather than the parent) so I cannot use it for this situation. Is this the case? If so, what's the best work around? Creating a new class that inherits from StaticFW that has this parameter?

**Comment by montoyjh** [comment]
Thanks for bringing this up, I've run into this as well, because occasionally I want to calc_locs to be more specific for the elastic workflow as well.  Maybe PassCalcLocs should be refactored in a way that makes it possible to get calc locs for a specific firework, e. g. rather than pushing the calc_locs into a list, set a specific key or something.  Maybe the calc_loc argument could then be that key?  I kind of did this for PassStressStrain, having it set specific dictionary keys rather than pushing the stresses and strains to a list.

Also worth noting that I think we might refactor a lot of the PassX FWs that have proliferated, seems like they could probably fall under some more general structure.

**Comment by computron** [comment]
Hi all,

There are a couple of issues mixed in here, so let me try to address as much as I can:

>> occasionally I want to calc_locs to be more specific for the elastic workflow as well. Maybe PassCalcLocs should be refactored in a way that makes it possible to get calc locs for a specific firework, e. g. rather than pushing the calc_locs into a list, set a specific key or something. 
>> It's my understanding that calc_loc = True will grab the location of the most recently completed calculation (rather than the parent) so I cannot use it for this situation.

The calc_locs has two ways of getting previous calculation data via the CopyVaspOutputs Firetask:
(i) For calc_loc=True, it will look through the array of data in the "calc_locs" key in the fw_spec and give you back the last array entry. What this means practically is that if you have a single parent, calc_loc=True should will copy from the location of that single parent. If you have multiple parents, calc_loc=True will give you the location of the parent that completed most recently. Note that if you have some side-branch of calculations that finished more recently than your parent, you won't get information from that side branch. @blondegeek , does this resolve your concern?
(ii) For calc_loc = <some string> (rather than True), you will grab a specific calculation from the history of calculations. Note that all entries in the calc_locs key have a "name" parameter. This name parameter can be easily set when using the PassCalcLocs Firetask; the default used in most Firework implementations is the name of the Firework. But you can easily code a Firework that sets the name of entry in the calc_locs key to "structure A" or "Bob the calculation" in PassCalcLocs. Then, for the downstream CopyVaspOutputs firetasks (anything downstream), if you set calc_loc="Bob the calculation", it will keep going back in history until it encounters Bob. So this allows you to pick out a specific entry from the calc_locs.

>>  If so, what's the best work around? Creating a new class that inherits from StaticFW that has this parameter?

For @blondegeek , it's possible that the default implementation is just fine. Let me know if not.
For @montoyjh, the StaticFW currently doesn't let you use mode (ii) if grabbing a specific calc_loc by the name. This is easy enough to fix though. I'll do it now.

>> Also worth noting that I think we might refactor a lot of the PassX FWs that have proliferated, seems like they could probably fall under some more general structure.

Let's look into this separately, e.g. in a separate issue ticket or something. I am slowly making my way through reviewing all the atomate packages (as you may have noticed).

**Comment by computron** [comment]
@montoyjh I pushed b722084c9193230ac7eeb752131be18dc9c2ec9a to make the StaticFW more flexible. I'll do this also for the other miscellaneous VASP Fireworks as I go through the code review.

**Comment by blondegeek** [comment]
Hi Anubhav, 
Diagrammatically, I have 

```txt
 O1 O2
 |   |
 S1 S2
 \  /
  FW
```
where O1,O2 <- OptimizeFWs, S <- StaticFWs, FW = other Firework with both StaticFWs as parents.

So, if I'm understanding what you said, calc_loc = True should be sufficient for my needs because O1 and S1 are on a different "branch" from O2 and S2.

Thanks!

**Comment by computron** [comment]
Hi @blondegeek 

Yes, that is correct. With both S1 and S2 using calc_loc=True, S1 should always grab from O1 and S2 should always grab from O2.

However, when you get to "FW", that's when calc_loc=True will either give you S1 *or* S2, depending on which one you want. If you for some reason wanted to get some data from each of S1 and S2, you an still use calc_loc but using the String method (ii).

I'll close this for now, but anyone feel free to re-open as necessary.
